### PR TITLE
Cloud Workstations - Workstation Config

### PR DESCRIPTION
**Pull request by bschaatsbergen**
The leading pull request for this is: https://github.com/GoogleCloudPlatform/magic-modules/pull/7005, that should be merged first. (as I've split up the PRs for this new service)

If this PR is for Terraform, I acknowledge that I have:

- [x] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [x] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-the-terraform-providers), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [x] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/third_party/terraform/tests) (for handwritten resources or update tests).
- [x] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/main/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [x] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:new-resource
google_workstations_workstation_config
```

### CODE REVIEW

**Comment by bschaatsbergen** [comment]
@SarahFrench / @ScottSuarez 

I seem to face an issue with the `PersistentDirectories` property.
According to the API spec it's an array with objects of type 'PersistentDirectory'.

https://cloud.google.com/workstations/docs/reference/rest/v1beta/projects.locations.workstationClusters.workstationConfigs

I bump into this issue: 
```
│ Error: Error updating WorkstationConfig "projects/playground-bruno/locations/us-central1/workstationClusters/tf-test-workstation-cluster/workstationConfigs/tf-test-workstation-config": googleapi: Error 400: Invalid JSON payload received. Unknown name "persistentDirectory" at 'workstation_config.persistent_directories[0]': Cannot find field.
│ Details:
│ [
│   {
│     "@type": "type.googleapis.com/google.rpc.BadRequest",
│     "fieldViolations": [
│       {
│         "description": "Invalid JSON payload received. Unknown name \"persistentDirectory\" at 'workstation_config.persistent_directories[0]': Cannot find field.",
│         "field": "workstation_config.persistent_directories[0]"
│       }
│     ]
│   }
│ ]
```

I believe this is because I created a structure as following:

```hcl
  persistent_directories {
    persistent_directory {
      mount_path = "/home/app"
      gce_pd {
        size_gb = 30
      }
    }
  }
 ```
 
 I don't think that I need the `persistent_directory` attribute, as it's now trying to send that over the API. What's the best way to solve this?

**Comment by bschaatsbergen** [comment]
@rileykarson / @SarahFrench do you have any idea why this happens?
```
TF_ACC=1 TF_SCHEMA_PANIC_ON_ERROR=1 go test ./google-beta -v -run=TestAccWorkstationsWorkstationConfig_workstationConfigBasicExample -timeout 240m -ldflags="-X=github.com/hashicorp/terraform-provider-google-beta/version.ProviderVersion=acc"
# github.com/hashicorp/terraform-provider-google-beta/google-beta [github.com/hashicorp/terraform-provider-google-beta/google-beta.test]
google-beta/resource_workstations_workstation_config_test.go:22:13: undefined: testAccWorkstationsWorkstationConfig_WorkstationConfigBasicExample
FAIL	github.com/hashicorp/terraform-provider-google-beta/google-beta [build failed]
FAIL
```

**Comment by SarahFrench** [comment]
👋 Hi @bschaatsbergen  - I took a look at [the Diffs](https://github.com/GoogleCloudPlatform/magic-modules/pull/7017#issuecomment-1413832414) in the beta version of the provider and saw [this line which seems to be the cause of the issue](https://github.com/modular-magician/terraform-provider-google-beta/blob/a1eaae5822b6cc13305fe4acf381ecadde6f1893/google-beta/resource_workstations_workstation_config_test.go#L43)

```
func testAccorkstationsWorkstationConfig_update(context map[string]interface{}) string {
```

instead of the expected testAcc**W**orkstationsWorkstationConfig_WorkstationConfigBasicExample

I'll take a look to try and figure out what's preventing that W being present - it's a bit odd! 🤔

File: mmv1/third_party/terraform/tests/resource_workstations_workstation_config_test.go.erb
Comment (SarahFrench) [inline mmv1/third_party/terraform/tests/resource_workstations_workstation_config_test.go.erb:11]: Ah it's here! Just fix up the typo and it should work

File: mmv1/third_party/terraform/tests/resource_workstations_workstation_config_test.go.erb
Comment (SarahFrench) [inline mmv1/third_party/terraform/tests/resource_workstations_workstation_config_test.go.erb:33]: And here

File: mmv1/third_party/terraform/tests/resource_workstations_workstation_config_test.go.erb
Comment (bschaatsbergen) [reply mmv1/third_party/terraform/tests/resource_workstations_workstation_config_test.go.erb:11]: Fixed

File: mmv1/third_party/terraform/tests/resource_workstations_workstation_config_test.go.erb
Comment (bschaatsbergen) [reply mmv1/third_party/terraform/tests/resource_workstations_workstation_config_test.go.erb:33]: Fixed

File: mmv1/third_party/terraform/tests/resource_workstations_workstation_config_test.go.erb
Comment (bschaatsbergen) [reply mmv1/third_party/terraform/tests/resource_workstations_workstation_config_test.go.erb:45]: Fixed

**Comment by bschaatsbergen** [comment]
> 👋 Hi @bschaatsbergen - I jumped to conclusions a bit when replying but I identified some typos that would be good to clean up!

Haha, how could I! Thanks for spotting those - I've fixed them. Though I still run into the above mentioned issue..

**Comment by SarahFrench** [comment]
Take # 2 of addressing your original question is that in [the code here](https://github.com/GoogleCloudPlatform/magic-modules/blob/101a2f6ca728d83378ce483f02bd8474893e6d4d/mmv1/third_party/terraform/tests/resource_workstations_workstation_config_test.go.erb#L24) you're referring to a function that would usually be produced by Magic Modules, but MM isn't generating tests.

To make MM create `testAccWorkstationsWorkstationConfig_WorkstationConfigBasicExample` you'd need to add an `examples` section to this new resource in the terraform.yaml file similar to this example for data stream resources: https://github.com/GoogleCloudPlatform/magic-modules/blob/b2719d82a46293604da993be231a5541500aeaea/mmv1/products/datastream/terraform.yaml#L52-L81

The new `examples` section would then point to your `mmv1/templates/terraform/examples/workstation_config_basic.tf.erb` file and use it to generate a test. The `name` attribute in these blocks are referring to files in `mmv1/templates/terraform/examples`, so for example the first example in the code posted above is using `mmv1/templates/terraform/examples/datastream_connection_profile_basic.tf.erb` to generate [one of the tests in this file](https://github.com/modular-magician/terraform-provider-google/blob/1d3b3e6cfdb313cea87e6efcea2a6f136d12a992/google/resource_datastream_stream_generated_test.go).

Does this help?

**Comment by bschaatsbergen** [comment]
> Take # 2 of addressing your original question is that in [the code here](https://github.com/GoogleCloudPlatform/magic-modules/blob/101a2f6ca728d83378ce483f02bd8474893e6d4d/mmv1/third_party/terraform/tests/resource_workstations_workstation_config_test.go.erb#L24) you're referring to a function that would usually be produced by Magic Modules, but MM isn't generating tests.
> 
> 
> 
> To make MM create `testAccWorkstationsWorkstationConfig_WorkstationConfigBasicExample` you'd need to add an `examples` section to this new resource in the terraform.yaml file similar to this example for data stream resources: https://github.com/GoogleCloudPlatform/magic-modules/blob/b2719d82a46293604da993be231a5541500aeaea/mmv1/products/datastream/terraform.yaml#L52-L81
> 
> 
> 
> The new `examples` section would then point to your `mmv1/templates/terraform/examples/workstation_config_basic.tf.erb` file and use it to generate a test. The `name` attribute in these blocks are referring to files in `mmv1/templates/terraform/examples`, so for example the first example in the code posted above is using `mmv1/templates/terraform/examples/datastream_connection_profile_basic.tf.erb` to generate [one of the tests in this file](https://github.com/modular-magician/terraform-provider-google/blob/1d3b3e6cfdb313cea87e6efcea2a6f136d12a992/google/resource_datastream_stream_generated_test.go).
> 
> 
> 
> Does this help?

Thanks for the extensive comment Sarah, I actually have added an example.

File: mmv1/third_party/terraform/tests/resource_workstations_workstation_config_test.go.erb
Comment (rileykarson) [inline mmv1/third_party/terraform/tests/resource_workstations_workstation_config_test.go.erb:24]:
```suggestion
				Config: testAccWorkstationsWorkstationConfig_workstationConfigBasicExample(context),
```

I think this might be a casing issue? https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-7017-old..auto-pr-7017#diff-594d28d38e532db3d2f0d1711c97ce2b2d75144a43731f3d58a003b91b952ae9R51

Comment (bschaatsbergen) [reply mmv1/third_party/terraform/tests/resource_workstations_workstation_config_test.go.erb:24]: Sharp riley!

Comment (SarahFrench) [reply mmv1/third_party/terraform/tests/resource_workstations_workstation_config_test.go.erb]: I'm definitely juggling one to many things - I searched the referenced function and saw it wasn't present in the downstream and assumed - my bad

**Comment by bschaatsbergen** [comment]
@rileykarson / @SarahFrench could you share the error with me regarding the tests? all tests seem to run fine on my local machine.

**Comment by SarahFrench** [comment]
Here's the error from the first test, but it's the same error across all of the workstation config test failures:

```
=== RUN   TestAccWorkstationsWorkstationConfig_update
=== PAUSE TestAccWorkstationsWorkstationConfig_update
=== CONT  TestAccWorkstationsWorkstationConfig_update
    provider_test.go:315: Step 1/4 error: Error running apply: exit status 1
        
        Error: Error waiting to create WorkstationCluster: Error waiting for Creating WorkstationCluster: Error code 13, message: an internal error has occurred
        
          with google_workstations_workstation_cluster.default,
          on terraform_plugin_test.tf line 21, in resource "google_workstations_workstation_cluster" "default":
          21: resource "google_workstations_workstation_cluster" "default" {
        
--- FAIL: TestAccWorkstationsWorkstationConfig_update (3220.56s)
FAIL
FAIL	github.com/hashicorp/terraform-provider-google-beta/google-beta	3220.607s
FAIL
```

**Comment by bschaatsbergen** [comment]
> Here's the error from the first test, but it's the same error across all of the workstation config test failures:
> 
> ```
> === RUN   TestAccWorkstationsWorkstationConfig_update
> === PAUSE TestAccWorkstationsWorkstationConfig_update
> === CONT  TestAccWorkstationsWorkstationConfig_update
>     provider_test.go:315: Step 1/4 error: Error running apply: exit status 1
>         
>         Error: Error waiting to create WorkstationCluster: Error waiting for Creating WorkstationCluster: Error code 13, message: an internal error has occurred
>         
>           with google_workstations_workstation_cluster.default,
>           on terraform_plugin_test.tf line 21, in resource "google_workstations_workstation_cluster" "default":
>           21: resource "google_workstations_workstation_cluster" "default" {
>         
> --- FAIL: TestAccWorkstationsWorkstationConfig_update (3220.56s)
> FAIL
> FAIL	github.com/hashicorp/terraform-provider-google-beta/google-beta	3220.607s
> FAIL
> ```
> 
> Perhaps the difference is that the tests are running in parallel, versus perhaps you were running tests once at a time?

That's right, I ran them once a time 👍🏼

**Comment by SarahFrench** [comment]
I'm currently running the tests in parallel against a GCP project I have access to and I've experienced a similar 'code 13' internal error at a different step (update instead of create).

Potentially we could
- create a permanent workstation cluster that is reused between workstation config tests
- or, make the tests not run in parallel if we find that definitely helps avoid errors

I'm not sure what the preferred solution is, so I'll go and get some input from others. I suspect we wouldn't want the tests to run sequentially as the clusters take a long time to create and delete

**Comment by SarahFrench** [comment]
Hi again @bschaatsbergen - I talked with others about this problem and there was a suggestion to reduce the number of acceptance tests running in parallel by creating one handwritten test that checks multiple update operations sequentially. The end result would be more test steps in the [[]resource.TestStep slice](https://github.com/GoogleCloudPlatform/magic-modules/blob/6140a5acab4581f8c772e398860ba0f494275132/mmv1/third_party/terraform/tests/resource_workstations_workstation_config_test.go.erb#L22-L40) of this test, using code that you copy over from the generated tests ([resource_workstations_workstation_config_generated_test.go](https://github.com/modular-magician/terraform-provider-google-beta/blob/97aea2174754a9dacf707385b2611c834fab2e06/google-beta/resource_workstations_workstation_config_generated_test.go)) into your handwritten test file.

To stop the generated tests being produced, you would need to add `skip_test: true` to the yaml defining the examples you made.

Does that make sense? Hopefully it'll address the failures, however it's still not 100% clear what conditions trigger the internal error.

**Comment by bschaatsbergen** [comment]
@SarahFrench - thanks a lot for your extensive reply, I'll be working on what you've addressed in your last comment!

**Comment by bschaatsbergen** [comment]
I expect this test to run very long, as some tests might cause a recreate to happen - @SarahFrench. But lets see what comes out of the VCR test 👍🏼

**Comment by bschaatsbergen** [comment]
Could you please share why the diff gen fails @SarahFrench

**Comment by bschaatsbergen** [comment]
@SarahFrench can you share details about the above please 🥲

**Comment by SarahFrench** [comment]
It's not over yet - the acceptance tests are initially run in an environment where 'recorded' API responses are fed to tests. If a test wants to send an outgoing request to an API which doesn't match recorded behaviour then the test fails and then there's a second stage where the test is run and is allowed to communicate directly with the Google APIs it needs.

The posts that say `Tests failed during RECORDING mode` are the ones to look out for!

**Comment by bschaatsbergen** [comment]
Sweet! They passed @SarahFrench woohoo

**Comment by bschaatsbergen** [comment]
Locally I get the same internal error..

File: mmv1/third_party/terraform/tests/resource_workstations_workstation_config_test.go.erb
Comment (SarahFrench) [inline mmv1/third_party/terraform/tests/resource_workstations_workstation_config_test.go.erb:73]:
```suggestion
	"foo" = "bar"
```

File: mmv1/third_party/terraform/tests/resource_workstations_workstation_config_test.go.erb
Comment (SarahFrench) [inline mmv1/third_party/terraform/tests/resource_workstations_workstation_config_test.go.erb:90]:
```suggestion
	"foo" = "bar"
```

**Review by SarahFrench — COMMENTED** [review]
I got some information that should help with this PR - the internal error could be linked to having a label called `update`.

I've made this change locally, using a commit from when you had multiple tests, and I ran the tests in parallel. Twice this was the outcome, with the `_update` test failing both times and all else passing:

```
--- PASS: TestAccWorkstationsWorkstationConfig_workstationConfigBasicExample

--- PASS: TestAccWorkstationsWorkstationConfig_workstationConfigContainerExample

--- PASS: TestAccWorkstationsWorkstationConfig_workstationConfigPersistentDirectoriesExample

--- PASS: TestAccWorkstationsWorkstationConfig_workstationConfigShieldedInstanceConfigExample

--- FAIL: TestAccWorkstationsWorkstationConfig_update
```

It might be worth reverting to the past commit when you had these tests and removing the `_update` test?

**Comment by bschaatsbergen** [comment]
> I got some information that should help with this PR - the internal error could be linked to having a label called `update`.
> 
> I've made this change locally, using a commit from when you had multiple tests, and I ran the tests in parallel. Twice this was the outcome, with the `_update` test failing both times and all else passing:
> 
> ```
> --- PASS: TestAccWorkstationsWorkstationConfig_workstationConfigBasicExample
> 
> --- PASS: TestAccWorkstationsWorkstationConfig_workstationConfigContainerExample
> 
> --- PASS: TestAccWorkstationsWorkstationConfig_workstationConfigPersistentDirectoriesExample
> 
> --- PASS: TestAccWorkstationsWorkstationConfig_workstationConfigShieldedInstanceConfigExample
> 
> --- FAIL: TestAccWorkstationsWorkstationConfig_update
> ```
> 
> It might be worth reverting to the past commit when you had these tests and removing the `_update` test?

Thanks for sharing this finding Sarah, I'll jump on it!

**Comment by bschaatsbergen** [comment]
@SarahFrench can we rerun the generate diffs and other tests? The update test seems to work now, I haven't had time to run the other tests today - but I expect that they work now.

**Comment by SarahFrench** [comment]
There was a change in this repo's CI - could you please rebase your branch so that [this commit](https://github.com/GoogleCloudPlatform/magic-modules/commit/d814900d4c535086cdb56de0e1480520d5271e53) from main is included in your branch?

**Comment by SarahFrench** [comment]
Also, just to manage expectations, I'm about to go on PTO and won't be able to review this PR again until Feb 20th. I'll keep an eye out for updates in the next 24 hours, but otherwise I won't be around to advise or merge if the tests pass.

**Comment by rileykarson** [comment]
@SarahFrench: You can hand the review off to me when you're unavailable!

**Comment by SarahFrench** [comment]
That sounds like a good idea - I'll remove myself as a reviewer now. Thanks @rileykarson !

**Comment by bschaatsbergen** [comment]
@SarahFrench have a good PTO, enjoy! Could you please see why the generate-diffs isn't running?

**Comment by rileykarson** [comment]
We require maintainer comments at the moment, sorry about that!

/gcbrun

**Comment by bschaatsbergen** [comment]
All good @rileykarson, thanks for triggering! We're getting there! :)

**Comment by bschaatsbergen** [comment]
I see that the gen diffs failed, I'm curious why - probably something related to my latest commit @rileykarson

**Review by rileykarson — COMMENTED** [review]
The error is `Psych::SyntaxError: (<unknown>): could not find expected ':' while scanning a simple key at line 416 column 15`. indicating we're running into a YAML parsing error. I don't see an obvious line where that's happening offhand! I'll comment if I spot it.

That _should_ be reproducible locally.

**Comment by bschaatsbergen** [comment]
I think that I fixed it just now in: [c5a6b7c](https://github.com/GoogleCloudPlatform/magic-modules/pull/7017/commits/c5a6b7c070f0cd2be44c1a6072cfb01b0f5e381a) @rileykarson

**Comment by bschaatsbergen** [comment]
All tests passed, great 👍🏼. Adding one more KMS key related test.

**Comment by bschaatsbergen** [comment]
Morning @rileykarson, if you get online could you trigger the checks for me - I suppose we can merge this today 🥳

**Comment by bschaatsbergen** [comment]
I believe /gcbrun is collaborator only? @rileykarson

**Comment by rileykarson** [comment]
Running generate-diffs! Generation is currently restricted.

/gcbrun

**Comment by bschaatsbergen** [comment]
Thanks @rileykarson, I'm heading off to bed - can you keep an eye out on the VCR results? TTYL 👍🏼

**Comment by bschaatsbergen** [comment]
Everything passed 👍🏼

**Review by rileykarson — COMMENTED** [review]
I've got some clarifications about a few of the optional fields! See inline.

File: mmv1/products/workstations/terraform.yaml
Comment (bschaatsbergen) [reply mmv1/products/workstations/terraform.yaml:42]: Yep!

File: mmv1/products/workstations/api.yaml
Comment (bschaatsbergen) [reply mmv1/products/workstations/api.yaml:224]: I've set required to `true`.

File: mmv1/products/workstations/api.yaml
Comment (bschaatsbergen) [reply mmv1/products/workstations/api.yaml:446]: Added

File: mmv1/products/workstations/api.yaml
Comment (bschaatsbergen) [reply mmv1/products/workstations/api.yaml:349]: Added a test: https://github.com/GoogleCloudPlatform/magic-modules/pull/7017/commits/0f7d6c58f9059a7f58ede907fea4847ee6317741

File: mmv1/products/workstations/api.yaml
Comment (bschaatsbergen) [reply mmv1/products/workstations/api.yaml:264]: Added a test: https://github.com/GoogleCloudPlatform/magic-modules/pull/7017/commits/0f7d6c58f9059a7f58ede907fea4847ee6317741

File: mmv1/products/workstations/api.yaml
Comment (bschaatsbergen) [reply mmv1/products/workstations/api.yaml:270]: Added a test that completely omits the host field, is that sufficient, or should we have another test that declares an empty 'host' attribute block? I don't think that's needed as the API sets the gceInstance by default in that case, same if we omit the entire `host` attribute block.

**Comment by bschaatsbergen** [comment]
Can you trigger the checks please @rileykarson

**Comment by bschaatsbergen** [comment]
Hey @rileykarson can we get this reviewed again :)?

**Comment by mar-pan** [comment]
Hi guys how this PR is doing?

**Comment by bschaatsbergen** [comment]
> Hi guys how this PR is doing?

Ready to be merged, just waiting for a review

**Review by rileykarson — COMMENTED** [review]
Sorry about the delay on my end! I lost a fair number of cycles this week dealing with the bug that prompted https://github.com/hashicorp/terraform-provider-google/releases/tag/v4.53.1 and have been behind on reviews since.

/gcbrun

**Comment by bschaatsbergen** [comment]
Can you share if the jobs failed or did they not ran yet?

**Comment by rileykarson** [comment]
Aha! It needs a _comment_, but what I did before was leave a _review comment_ which is different.

**Comment by bschaatsbergen** [comment]
Ship it! @rileykarson

File: mmv1/third_party/terraform/tests/resource_workstations_workstation_config_test.go.erb
Comment (SarahFrench) [inline mmv1/third_party/terraform/tests/resource_workstations_workstation_config_test.go.erb:8]: A solution to the erroring check on this PR would be to change this file to have a single set of `<% unless version == "ga" -%>` and `<% end -%>` that contains all the code _except_ `package google`. In the GA provider the file would only contain `package google` and in the beta provider the full file would be present.

**Review by SarahFrench — COMMENTED** [review]
👋  I can help with info about the check failures. The run-rake-tests check isn't relevant to this PR and is due to upgrades happening in our CI. The other failing test produces this error when testing the GA provider code generated from this PR:

```
google/resource_workstations_workstation_config_test.go:4:2: imported and not used: "testing"
google/resource_workstations_workstation_config_test.go:6:2: imported and not used: "github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
```

This is because `<% unless version == "ga" -%>` prevents test code being included in the GA provider, and the file is left with unused imports. See other comment for a suggested solution!

File: mmv1/third_party/terraform/tests/resource_workstations_workstation_config_test.go.erb
Comment (bschaatsbergen) [reply mmv1/third_party/terraform/tests/resource_workstations_workstation_config_test.go.erb:8]: Hey @SarahFrench, thanks for the suggestion. I'll apply it straight away. Hope you had a nice time off, welcome back :).

Comment (bschaatsbergen) [reply mmv1/third_party/terraform/tests/resource_workstations_workstation_config_test.go.erb:9]: I've just moved the `<% unless version == "ga" -%>` back-up above the imports, I suppose this would be sufficient too? If so, could you please trigger the /gcbrun :)

Comment (SarahFrench) [reply mmv1/third_party/terraform/tests/resource_workstations_workstation_config_test.go.erb:9]: Will do!

**Comment by bschaatsbergen** [comment]
@SarahFrench did the rake test fail again or did it not run yet?

**Comment by bschaatsbergen** [comment]
> @SarahFrench did the rake test fail again or did it not run yet?

I've updated the code to make it a single set of `unless ga/end` as you suggested.

**Comment by SarahFrench** [comment]
>@SarahFrench did the rake test fail again or did it not run yet?

The rake test did fail, but it's not linked to your PR in any way and wouldn't block us merging this PR

**Comment by bschaatsbergen** [comment]
@SarahFrench woohoo, looks it first complained about one test (during vcr recording). But all seems passed now. If we can get this merged I'm going to jump on the last Cloud Workstations resource!

**Comment by bschaatsbergen** [comment]
@SarahFrench woohoo, looks it first complained about one test (during vcr recording). But all seems passed now. If we can get this merged I'm going to jump on the last Cloud Workstations resource! #7018

Cc @rileykarson

**Comment by rileykarson** [comment]
Handing back to @SarahFrench, sorry for the delays on my end!

**Comment by bschaatsbergen** [comment]
> Handing back to @SarahFrench, sorry for the delays on my end!

All good, thanks for your involvement @rileykarson.

File: mmv1/products/workstations/terraform.yaml
Comment (SarahFrench) [inline mmv1/products/workstations/terraform.yaml:41]:
```suggestion
```
This line should be deleted from the array, as it causes the documentation to say you can import a workstation config with the id of a workstation cluster ([here's a preview of how the docs would be generated from the PR currently that shows this](https://github.com/modular-magician/terraform-provider-google-beta/blob/13b7d73767b3ff56fdc482ef6e295e23b24021c6/website/docs/r/workstations_workstation_config.html.markdown#import))

File: mmv1/templates/terraform/examples/workstation_config_basic.tf.erb
Comment (SarahFrench) [inline mmv1/templates/terraform/examples/workstation_config_basic.tf.erb:3]: Are these data sources used in the examples? If not, I think you should remove them from the examples to avoid confusion and reduce the size of the example blocks [in the generated documentation](https://github.com/modular-magician/terraform-provider-google-beta/blob/13b7d73767b3ff56fdc482ef6e295e23b24021c6/website/docs/r/workstations_workstation_config.html.markdown).

**Review by SarahFrench — COMMENTED** [review]
I've had a read through and I found a few things that can be improved, but they're all small. Sorry for the delay reviewing - but we're definitely getting close now!

**Comment by bschaatsbergen** [comment]
Thanks @SarahFrench I'm home in a couple of hours and will get it sorted.

File: mmv1/templates/terraform/examples/workstation_config_basic.tf.erb
Comment (bschaatsbergen) [reply mmv1/templates/terraform/examples/workstation_config_basic.tf.erb]: Indeed not used, removed it.

File: mmv1/products/workstations/terraform.yaml
Comment (bschaatsbergen) [reply mmv1/products/workstations/terraform.yaml]: Removed this line.

File: mmv1/products/workstations/api.yaml
Comment (bschaatsbergen) [reply mmv1/products/workstations/api.yaml]: removed duplicate output entry

File: mmv1/products/workstations/api.yaml
Comment (bschaatsbergen) [reply mmv1/products/workstations/api.yaml]: I've removed this output and created a test 👍🏼

**Comment by bschaatsbergen** [comment]
Heya @SarahFrench, I've addressed the comments :)!

**Comment by bschaatsbergen** [comment]
I believe that I messed something up in the test, moved the updating of the displayName field in the test through the context now. Can we re-run the checks? @SarahFrench

**Comment by SarahFrench** [comment]
That commit looks ok to me and should run. Mutating the value in context feels like a bad idea generally, but in this case it's a very simple test so I think it's ok.

**Comment by bschaatsbergen** [comment]
> 05ef367

I felt the same, at first I was injecting the value and interpolating it in the terraform code - but I saw a lot of tests that do it on a context level. So wasn't quite sure what the standard was.

**Comment by SarahFrench** [comment]
> wasn't quite sure what the standard was 

I think for handwritten update tests the typical way is to define two separate functions (with names ending in `_basic` or `_update`) for the 'before' and 'after' config. Any updates in the `_update` function's config are hardcoded, e.g. `display_name = "%{display_name} %{random_suffix}"` in `_basic` versus `display_name = "%{display_name} %{random_suffix}-update"` in `_update`, and the context map stays the same.

But having said that, if you're making a very small change then making 1 reusable function instead of 2 makes sense. 🤷 In this case I think mutating context is ok but only because the test is so simple and the field being mutated isn't used for identifying resources via the API.

**Comment by bschaatsbergen** [comment]
@SarahFrench 🥳 
> All tests passed

**Review by SarahFrench — APPROVED** [review]
Ok let's get this merged! :D

Good luck with implementing the 3rd resource for https://github.com/hashicorp/terraform-provider-google/issues/12763

**Comment by bschaatsbergen** [comment]
Woohoo! Amazing, thanks for the extensive review @SarahFrench - on to the last resource!

**Pull request merged**
